### DISCUSSION TITLE
What HA Automations are you all using?

### DISCUSSION THREAD

**Discussion by NickSutton** [BottlecapDave/HomeAssistant-OctopusEnergy#530] [General]
In order to prevent EV charging emptying my solar battery storage, I also need to set my inverter in to 'charging' mode when an off-peak period lands, it's not a problem it's pretty handy.

That said, I am repeatedly caught out because my automation is only looking at the first 'start' time and last 'end' time to calculate the charging period. 

Hi All,

I've been using [this Intelligent Octopus HACS integration](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/0959223aafe951af591b38154541438d7c9cfce9/_docs/entities/electricity.md#home-mini-entities) to get information about by off peak rates when my EV is plugged in. It works great and I'm really happy with it.

It turns out the that the OI integration sends an array of planned charging periods like so:

```
[ 
    { start: '2023-11-06T00:00:00+00:00', end: '2023-11-06T00:30:00+00:00' },
    { start: '2023-11-06T01:00:00+00:00', end: '2023-11-06T01:30:00+00:00' },
    { start: '2023-11-06T00:30:00+00:00', end: '2023-11-06T01:00:00+00:00' },
    { start: '2023-11-06T02:00:00+00:00', end: '2023-11-06T02:30:00+00:00' },
    { start: '2023-11-06T01:30:00+00:00', end: '2023-11-06T02:00:00+00:00' },
    { start: '2023-11-06T03:00:00+00:00', end: '2023-11-06T03:30:00+00:00' }
]
```

I am using a template to grab the first `start` time and last `end` time of the array. 
My automation to write the charging period to the inverter requires only `HH:MM`

However, there are a number of scenarios I've overlooked and need some help to work out how I can reliably sync the inverter charging with off-peak windows from IO:

1 - Fixed off-peak is 23:30 - 05:30 every night, regardless of EV charging etc... If the EV only needs a small charge, when plugged in IO comes back with something like 0100 - 0400, which means I've lost out on a couple of hours of off-peak charging the solar batteries (as 0100-0400 would be written to the inverter.  What would be the best way to ensure 2330-0530 is respected if the window falls within those times?

2 - The charging periods are not always concurrent. For example, I was home early today and plugged in with a nearly flat battery. The inverter programmed itself to charge from 1300 - 0700 the next day! However, on closer inspection there were several gaps in the periods (where demand on the grid is high). My logic does not account for this...

Surely someone has trodden this path before I? I would be really interested to hear what approaches others are taking.

Many thanks

**Comment by BottlecapDave** [comment]
Hi. To my knowledge a few people take advantage of the [target rate sensors](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md) along with the [update target rate config service](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/services.md#service-octopus_energyupdate_target_config). Theses sensors find the best continuous or intermittent times based on rates. They then use this service to adjust the target hours based on how much they need to charge their battery (e.g 80-90% = 1 hour). This sensor exposes the discovered times via one of the attributes so you can use this to set the times for you battery.

This is obviously just one way of doing things :)

**Reply by NickSutton** [reply]
Thanks, I'll read up on those. I didn't think I needed them
